Project Manager Gem Version Support

AMZN-alexpete opened the pull request:
## What does this PR do?

This gem will need an exception because it adds new functionality - I'm in the process of requesting that exception, but wanted to get the review started.

- add Gem's compatible engines data to inspector in Gem Catalog
- enables users to select which version of a gem to use in their project
![image](https://user-images.githubusercontent.com/26804013/229592229-7ebeb6b4-bab5-48d3-8971-8977d74a66d5.png)

- fix not showing missing gems a project uses in the Gem Catalog
![image](https://user-images.githubusercontent.com/26804013/228114357-dabb99b7-8afe-4c15-acf7-a4780c99e85b.png)

- fixes inability to deactivate gems in a project that are not registered or are missing
- fixes crash in shutdown when disconnect from download signal
- fixes Gem download of GitHub archives failing to register gem in subfolder
- fix compatibility check using old version of gems in dependency check and failing
- fix Project Mananger gem compatibility check using optional gems and failing when not found
- fix requirements showing for gems that are already added - but still shows requirements when gem version is changed
- fix a few styling issues in Gem Catalog inspector
- refactors `GemModel` to remove unnecessary Qt Roles 
- removes unused `GetEngineGems` function
- fixes https://github.com/o3de/o3de/issues/15557


## How was this PR tested?

- ran o3de pytests and Project Manager unit tests
- added missing gems to a project and verified it appeared in Gem Catalog for that project and could be removed
- added multiple versions of a gem and verified the gem version could be changed
- downloaded a remote gem from GitHub and verified it could be activated in the project

lumberyard-employee-dm commented on Code/Tools/ProjectManager/Source/GemCatalog/GemItemDelegate.h:
```diff
@@ -75,6 +75,10 @@ namespace O3DE::ProjectManager
         inline constexpr static int s_platformTextLineBottomMargin = 5;
         inline constexpr static int s_platformTextWrapAroundLineMaxCount = 2;
 
+        // Version
+        inline constexpr static int s_versionSize = 50;
+        inline constexpr static int s_versionSizeSpacing = 25;
```
The `static` keyword isn't needed.
internal linkage isn't needed for each constant.

AMZN-alexpete replied:
looks like inline cannot be used without 'static' and these vars are referred to in other .cpp files like `GemCatalogScreen.cpp` and `GemListView.cpp` .  I could make accessor functions, but I might leave that for another PR.

lumberyard-employee-dm replied:
Oh, these are static class members.
Never mind then.
I thought these were non-member constexpr constants.

lumberyard-employee-dm commented on Code/Tools/ProjectManager/Source/GemCatalog/GemRequirementFilterProxyModel.cpp:
```diff
@@ -24,7 +24,8 @@ namespace O3DE::ProjectManager
         QModelIndex sourceIndex = sourceModel()->index(sourceRow, 0, sourceParent);
-        return GemModel::IsAdded(sourceIndex) && GemModel::HasRequirement(sourceIndex);
+        const bool hasNewVersion = !GemModel::GetNewVersion(sourceIndex).isEmpty();
+        return GemModel::IsAdded(sourceIndex) && GemModel::HasRequirement(sourceIndex) && (!GemModel::WasPreviouslyAdded(sourceIndex) || hasNewVersion) ;
```
nitpick extra space before semicolon
```suggestion
        return GemModel::IsAdded(sourceIndex) && GemModel::HasRequirement(sourceIndex) && (!GemModel::WasPreviouslyAdded(sourceIndex) || hasNewVersion);
```

AMZN-alexpete commented on scripts/o3de/tests/test_disable_gem.py:
```diff
@@ -243,6 +246,9 @@ def is_file(path : pathlib.Path) -> bool:
             project_json = get_project_json_data(project_path=project_path)
             assert expected_gem_name in project_json.get('gem_names', [])
 
+            # unregister the gem if asked
+            self.gem_registered = not unregister_gem
```
My comments here could be more helpful for sure, it's supposed to be letting the test specify whether to assume the gem has already been registered or not so we can test cases when the gem isn't registered, but should still be disabled in a project.

tkothadev replied:
maybe renaming it as `self.assume_gem_registered` would make the intent clearer?

tkothadev commented on Code/Tools/ProjectManager/Source/GemCatalog/GemItemDelegate.cpp:
```diff
@@ -157,19 +156,29 @@ namespace O3DE::ProjectManager
+            QRect gemVersionRect{ versionXBounds.first, contentRect.top(), versionXBounds.second - versionXBounds.first, contentRect.height() };
+            painter->setFont(standardFont);
+            gemVersionRect = painter->boundingRect(gemVersionRect, Qt::TextWordWrap | Qt::AlignRight | Qt::AlignVCenter, gemInfo.m_version);
+            painter->drawText(gemVersionRect, Qt::TextWordWrap | Qt::AlignRight | Qt::AlignVCenter, gemInfo.m_version);
```
Does the word wrap occur only after each `.` of the semantic version?

AMZN-alexpete replied:
At word boundaries: https://doc.qt.io/qt-5/qt.html#TextFlag-enum

tkothadev reviewed (approved):
Changes look good to me. Had some minor questions.

lumberyard-employee-dm commented on Code/Tools/ProjectManager/Source/GemCatalog/GemModel.cpp:
```diff
@@ -27,53 +31,218 @@ namespace O3DE::ProjectManager
+            {
+                if (existingGemInfo.m_version == gemInfo.m_version)
+                {
+                    AZ_Info("ProjectManager", "Not adding GemInfo because a GemInfo with path (%s) and version (%s) already exists.",
```
Nope it does not, it uses Printf

AMZN-alexpete replied:
It looks like Printf adds a newline:
https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Debug/Trace.h#L114

lumberyard-employee-dm replied:
Actually it doesn't because that ITrace::Printf is overridden by the implementation class of [Trace::Printf](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Debug/Trace.cpp#L532-L559)

lumberyard-employee-dm commented on scripts/o3de/o3de/project_manager_interface.py:
```diff
@@ -268,18 +269,30 @@ def get_gem_info(gem_path: str) -> dict or None:
+    # flatten into a single list
+    all_gem_json_data = [gem_json_data for gem_versions in all_gem_json_data.values() for gem_json_data in gem_versions]
+
+    # it's easier to determine which gems are engine gems here rather than in c++
+    # because this gem might belong to a different engine than the one Project Manager is
+    # running out of
+    engine_path = manifest.get_project_engine_path(project_path=project_path) if project_path else manifest.get_this_engine_path() 
+    for i, gem_json_data in enumerate(all_gem_json_data):
+        all_gem_json_data[i]['engine_gem'] = engine_path in gem_json_data['path'].parents
```
Can't this check against the `manifest.get_engine_gems` gems?
```suggestion
    # it's easier to determine which gems are engine gems here rather than in c++
    # because this gem might belong to a different engine than the one Project Manager is
    # running out of
    engine_path = manifest.get_project_engine_path(project_path=project_path) if project_path else manifest.get_this_engine_path() 
    engine_gems = manifest.get_engine_gems(engine_path);
    for i, gem_json_data in enumerate(all_gem_json_data):
        all_gem_json_data[i]['engine_gem'] = gem_json_data['path'] in engine_gems
```

AMZN-alexpete replied:
I found at least one case where the drive letter case was different leading to invalid matches e.g. 'D:' vs 'd:' because the 'path' values in `gem_json_data` are strings and so are those in `engine_gems`. I think this is due to how you can have an `engine_path` entry in a `user/project.json` that points to an all lower case path where as the manifest.json may have an upper case version.
```
>>> import pathlib
>>> p = pathlib.Path('c:/Windows')
>>> p.as_posix()
'c:/Windows'

>>> p2 = pathlib.Path('C:/Windows')
>>> p2.as_posix()
'C:/Windows'

>>> p3 = pathlib.Path('c:/windows')
>>> p3.as_posix()
'c:/windows'
```

lumberyard-employee-dm replied:
That is why the pathlib.Path should be used for comparison.
Doing a string compare is incorrect and a path compare should occur.

AMZN-alexpete commented on Code/Tools/ProjectManager/Source/GemCatalog/GemCatalogHeaderWidget.cpp:
```diff
@@ -353,7 +347,32 @@ namespace O3DE::ProjectManager
         {
-            tags.push_back({ GemModel::GetDisplayName(modelIndex), GemModel::GetName(modelIndex) });
+            const GemInfo& gemInfo = GemModel::GetGemInfo(modelIndex);
+            if(gemInfo.m_isEngineGem)
```
You cannot change engine gem versions like external gems,  there is only one version of an engine gem available because it comes with the engine.  Currently all of our engine gems are only assumed compatible with the engine they come with.

You cannot, for example, select a different version of the Atom gem, and normally you don't want to, when you register your project with a new engine, that's when you address compatibility for all the engine gems, unlike remote/external gems that can be updated whenever and may support many engine versions.

AMZN-alexpete commented on scripts/o3de/tests/test_download.py:
```diff
@@ -57,7 +57,10 @@
-    "gems": ["http://o3derepo.org/TestGem", "C:/localrepo/TestLocalGem"],
+    "gems": [
+        "http://o3derepo.org/TestGem", 
+        "C:/localrepo/TestLocalGem"
```
The tests pass when using `file://C:/localrepo/TestLocalGemfile://` but I haven't tested making a repo and adding it in Project Manager.  I may do that when I get to some of the other bugs that were brought up around downloading, and when we add versioning support for remote objects

lumberyard-employee-dm replied:
> `file://C:/localrepo/TestLocalGemfile://`

Was the final `file://` intended to be part of that comment?

AMZN-alexpete replied:
oh, sorry, copy-paste error on my part, I just tested `file://C:/localrepo/TestLocalGemfile`

AMZN-alexpete commented:
**Exception Tracking**
- [ ] 1. Owning SIG requests the exception, by a Sig Representative (Chair/co-chair) adding the PR to the exception requests queue board.
- [ ] 2. (@lsemp3d , SIG-Content) Owning SIG comments on the PR explaining impact if this code does not go into stabilization.
- [ ] 3. (@lumberyard-employee-dm , SIG-core) Unrelated SIG representative (chair/cochair) approves the exception, adding a comment indicating approval and the SIG they represent.
- [ ] 4. UX SIG representative (chair/cochair) approves the exception, adding a comment indicating approval and the SIG they represent.
- [ ] 5. Testing SIG representative (chair/cochair) approves the exception, adding a comment indicating approval and the SIG they represent.
- [x] (Must occur at some point before SIG-Release approval) Code has been approved for merging by maintainers
- [ ] 6. SIG-Release Release Manager or co-Release Manager approves the exception, adding a comment indicating approval and the SIG they represent.
- [ ] 7. If all approvals are accepted, then the exception PR may be merged
- [ ] 8. Merged PR is verified successful via testing. If Merged PR verification testing failed, consult reps from step 3-5 and discussion occurs on fixing the failure vs backing out change
- [ ] 9. Exception is complete

lsemp3d commented:
As SIG-Content chair I approve this to go into stabilization, it fixes many issues that stem from using an incorrect gem version

lumberyard-employee-dm commented:
> **Exception Tracking**
> 
>     * [x]  1. Owning SIG requests the exception, by a Sig Representative (Chair/co-chair) adding the PR to the exception requests queue board.
> 
>     * [x]  2. (@lsemp3d , SIG-Content) Owning SIG comments on the PR explaining impact if this code does not go into stabilization.
> 
>     * [x]  3. (@lumberyard-employee-dm , SIG-core) Unrelated SIG representative (chair/cochair) approves the exception, adding a comment indicating approval and the SIG they represent.
> 
>     * [ ]  4. UX SIG representative (chair/cochair) approves the exception, adding a comment indicating approval and the SIG they represent.
> 
>     * [ ]  5. Testing SIG representative (chair/cochair) approves the exception, adding a comment indicating approval and the SIG they represent.
> 
>     * [x]  (Must occur at some point before SIG-Release approval) Code has been approved for merging by maintainers
> 
>     * [ ]  6. SIG-Release Release Manager or co-Release Manager approves the exception, adding a comment indicating approval and the SIG they represent.
> 
>     * [ ]  7. If all approvals are accepted, then the exception PR may be merged
> 
>     * [ ]  8. Merged PR is verified successful via testing. If Merged PR verification testing failed, consult reps from step 3-5 and discussion occurs on fixing the failure vs backing out change
> 
>     * [ ]  9. Exception is complete

As the SIG-Core chair, I approve of this change to go into stabilization/2305 as an unrelated SIG representative

vincent6767 commented:
As the co-chair of SIG-Release, I approve of adding this change into stabilization.

rainbj commented:
![2023-03-29_10-07-14](https://user-images.githubusercontent.com/82551958/228623484-5e9f9037-1ab2-4bc4-a4d2-c157c05f39e4.png)

Hi Everyone. So I wanted to point out a few issues I saw that feel a bit confusing for me. The numbers below are referenced in the image above. 

1) Title on the modal calls out that I'm on the Manual setup step. Then the first sentence references that I still have to do manual setup before the project can be built. But all of this is not an action, nor does it tell a user anything about why I am here. What does pressing continue actually mean? To me, this entire pop-up has no value unless you are asking them to just know you are doing something... Which seems obvious that your doing something.  Unless we can make this seem clear about the purpose of what happens when I click continue, I suggest removing the modal all together.

2) IF we are keeping this window. Please remove all the blank space in the dark gray area (5-10px max around). Then make sure the sentence is on one line and not two. I assume this is supposed to show a "from version" and a "to version", but it just says multiple versions which is not useful.  This will always be true when upgrading a gem so saying multiple has no value. We need to be more specific or remove this text all together.

3) Showing this warning does help the user make a decision. It just looks like the Gem is broken, or it has an error. I would change the text to something like "New gem version available." with a new icons like this.
![2023-03-29_11-12-45](https://user-images.githubusercontent.com/82551958/228630189-31ceb100-1355-44af-be33-0fda39027b12.png) then change the button to "update now".

4) Additional information section has two values. Both say unknown, "**known date**" and "**unknown size**". This text just makes more noise and is not helpful to the customer. Maybe just remove this section when the property value comes back as unknown. 

5) I don't understand the difference between **Update** and **Edit**. Wouldn't an update be an edit? Couldn't I edit it to update it? Is this redundant? 

6) You have text in the right panel that reads " **An example of a gem that has multiple version**". I assume this is just temp text, so  you should remove it before releasing this. 

7) Lastly: "**Engine: o3de >= 1.1.0**" is confusing. I assume this means. **Engine Compatibility: 1.0 - 1.1.0 / All / 1.0 - Current**

AMZN-alexpete commented:
> Hi Everyone. So I wanted to point out a few issues I saw that feel a bit confusing for me. The numbers below are referenced in the image above.

Thanks for the feedback!
#1 #2  and #3 are not related to this change, it's the dialog that pops up when a gem has requirements and is not related to my changes.  I just happened to use a gem with requirements for my testing.  Gems that do not have requirements will not cause this popup.


> 4. Additional information section has two values. Both say unknown, "**known date**" and "**unknown size**". This text just makes more noise and is not helpful to the customer. Maybe just remove this section when the property value comes back as unknown.
I've considered removing this text, but it's primarily text that I just didn't update in my test gem.  For someone who is vending a gem, I expect they will fill in this information and not leave it as "unknown", but we can also add code to hide those fields if the text contains "unknown" too.

> 5. I don't understand the difference between **Update** and **Edit**. Wouldn't an update be an edit? Couldn't I edit it to update it? Is this redundant?

Update will attempt to pull down the latest version of the gem - it's old functionality that hasn't been changed in this PR
Edit takes you to edit the current gem information, also functionality that isn't changed by this PR

> 6. You have text in the right panel that reads " **An example of a gem that has multiple version**". I assume this is just temp text, so  you should remove it before releasing this.

Yup, it's just the text from my testing gem and will not be part of any release

> 7. Lastly: "**Engine: o3de >= 1.1.0**" is confusing. I assume this means. **Engine Compatibility: 1.0 - 1.1.0 / All / 1.0 - Current**

That's a good call out, I'd like to have a field that states gives you a warning if the gem is not known to be compatible with your project/engine, but for now I'm just showing any compatible engines that may be listed in the gem.json
o3de >= 1.1.0 means: This gem is compatible with any engine named o3de version 1.1.0 or higher.

AMZN-alexpete commented:
After talking with @rainbj  and @bhanuja-s I've made additional changes 
1. show an "update available" version filter
2. show a little yellow icon next to gems with available version updates
![image](https://user-images.githubusercontent.com/26804013/229219177-28b84ab8-ff26-4c0f-8213-47af6fd32756.png)

3. use the human readable compatible engine versions
4. use info icon instead of warning for requirements
5. show a cloud icon always for gems from remote repos
![image](https://user-images.githubusercontent.com/26804013/229218619-63dc9eed-79c0-4da4-b44d-5dffdad621e7.png)

And also
1. fixed missing gem size text
2. removed some debug qss
3. base if a gem is remote on if it has a valid repo url instead of the origin field

I still need to address one more UX request which is to show separate gem entries for remote vs local gems that have the same name so that if you have the remote version of PopcornFX and have cloned it on disk, there is a way to tell the difference in the Gem Catalog if possible.

bhanuja-s commented:
As the co-chair of SIG-UI-UX, I approve of adding this change into stabilization.

tonybalandiuk commented:
@AMZN-alexpete has this been tested post-merge?

AMZN-alexpete commented:
@tonybalandiuk yes, by me as part of the remote gem version support which was based on this.
